cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -1778,6 +1784,711 @@ class SortIndicesMetaFunction : public MetaFunction {
+        const auto rval = GetView::LogicalValue(arr.GetView(right));
+        return rval < lval;
+      };
+    }
+    arrow::internal::Heap<uint64_t, decltype(cmp)> heap(cmp);
+    uint64_t* iter = indices_begin;
+    for (; iter != kth_begin && heap.Size() < static_cast<size_t>(k_); ++iter) {
+      heap.Push(*iter);
+    }
+    for (; iter != end_iter && heap.Size() > 0; ++iter) {
+      uint64_t x_index = *iter;
+      const auto lval = GetView::LogicalValue(arr.GetView(x_index));
+      const auto rval = GetView::LogicalValue(arr.GetView(heap.Top()));
+      if (order_ == SortOrder::Ascending) {
+        if (lval < rval) {
+          heap.ReplaceTop(x_index);
+        }
+      } else {
+        if (rval < lval) {
+          heap.ReplaceTop(x_index);
+        }
+      }
+    }
+
+    int64_t out_size = static_cast<int64_t>(heap.Size());
+    ARROW_ASSIGN_OR_RAISE(
+        auto take_indices,
+        MakeMutableArrayForFixedSizedType(uint64(), out_size, ctx_->memory_pool()));
+
+    auto* out_cbegin = take_indices->GetMutableValues<uint64_t>(1) + out_size - 1;
+    while (heap.Size() > 0) {
+      *out_cbegin = heap.Top();
+      heap.Pop();
+      --out_cbegin;
+    }
+    ARROW_ASSIGN_OR_RAISE(*output_, Take(array_, Datum(std::move(take_indices)),
+                                         TakeOptions::NoBoundsCheck(), ctx_));
+    return Status::OK();
+  }
+
+  ExecContext* ctx_;
+  const Array& array_;
+  int64_t k_;
+  const std::shared_ptr<DataType> physical_type_;
+  SortOrder order_;
+  Datum* output_;
+};
+
+template <typename ArrayType>
+struct TypedHeapItem {
+  uint64_t index;
+  uint64_t offset;
+  ArrayType* array;
+};
+
+class ChunkedArraySelecter : public TypeVisitor {
+ public:
+  ChunkedArraySelecter(ExecContext* ctx, int64_t k, const ChunkedArray& chunked_array,
+                       const SortOrder order, Datum* output)
+      : TypeVisitor(),
+        k_(k),
+        chunked_array_(chunked_array),
+        physical_type_(GetPhysicalType(chunked_array.type())),
+        physical_chunks_(GetPhysicalChunks(chunked_array_, physical_type_)),
+        order_(order),
+        ctx_(ctx),
+        output_(output) {}
+
+  Status Run() { return physical_type_->Accept(this); }
+
+#define VISIT(TYPE) \
+  Status Visit(const TYPE& type) { return SelectKthInternal<TYPE>(); }
+
+  VISIT_PHYSICAL_TYPES(VISIT)
+
+#undef VISIT
+
+  template <typename InType>
+  Status SelectKthInternal() {
+    using GetView = GetViewType<InType>;
+    using ArrayType = typename TypeTraits<InType>::ArrayType;
+    using HeapItem = TypedHeapItem<ArrayType>;
+
+    const auto num_chunks = chunked_array_.num_chunks();
+    if (num_chunks == 0) {
+      return Status::OK();
+    }
+    if (k_ > chunked_array_.length()) {
+      k_ = chunked_array_.length();
+    }
+    std::function<bool(const HeapItem&, const HeapItem&)> cmp;
+    if (order_ == SortOrder::Ascending) {
+      cmp = [](const HeapItem& left, const HeapItem& right) -> bool {
```

Comment:
These ordering checks and compare lambdas can be resolved via templates based on the SelectionOperator (topK vs bottomK).